File: amazingcore/messages/common/player_maze.py
from amazingcore.messages.message_interfaces import SerializableMessage
from amazingcore.codec.bit_stream import BitStream
from amazingcore.messages.common.object_id import ObjectID
from amazingcore.messages.common.home_theme import HomeTheme

import datetime as dt


class PlayerMaze(SerializableMessage):
    def __init__(self,
                 aw_object_id: ObjectID = None,
                 name: str = None,
                 size: int = None,
                 thumbnail: bytes = None,
                 publish_timestamp: dt.datetime = None,
                 num_rooms: int = None,
                 num_tubes: int = None,
                 rating: int = None,
                 is_locked: bool = None,
                 is_home_maze: bool = None,
                 is_published: bool = None,
                 is_publish_expired: bool = None,
                 player_id: ObjectID = None,
                 maze_pieces: list = None,  # GSFPlayerMazePiece
                 home_theme: HomeTheme=None,
                 parent_id: ObjectID = None,
                 source_id: ObjectID = None):
        self.aw_object_id = aw_object_id
        self.name = name
        self.size = size
        self.thumbnail = thumbnail
        self.publish_timestamp = publish_timestamp
        self.num_rooms = num_rooms
        self.num_tubes = num_tubes
        self.rating = rating
        self.is_locked = is_locked
        self.is_home_maze = is_home_maze
        self.is_published = is_published
        self.is_publish_expired = is_publish_expired
        self.player_id = player_id
        self.maze_pieces = maze_pieces
        self.home_theme = home_theme
        self.parent_id = parent_id
        self.source_id = source_id

    def serialize(self, bit_stream: BitStream):
        bit_stream.write_start()
        self.aw_object_id.serialize(bit_stream)
        bit_stream.write_str(self.name)
        bit_stream.write_long(self.size)
        bit_stream.write_bytes(self.thumbnail)
        bit_stream.write_dt(self.publish_timestamp)
        bit_stream.write_short(self.num_rooms)
        bit_stream.write_short(self.num_tubes)
        bit_stream.write_short(self.rating, True)
        bit_stream.write_bool(self.is_locked)
        bit_stream.write_bool(self.is_home_maze)
        bit_stream.write_bool(self.is_published)
        bit_stream.write_bool(self.is_publish_expired)
        self.player_id.serialize(bit_stream)
        bit_stream.write_int(0)  # GSFPlayerMazePieces
        self.home_theme.serialize(bit_stream)
        self.parent_id.serialize(bit_stream)
        self.source_id.serialize(bit_stream)

    def deserialize(self, bit_stream: BitStream):
        raise NotImplementedError

    def to_dict(self):
        return {
            'aw_object_id': self.aw_object_id.to_dict(),
            'name': self.name,
            'size': self.size,
            'thumbnail': self.thumbnail,
            'publish_timestamp': self.publish_timestamp,
            'num_rooms': self.num_rooms,
            'num_tubes': self.num_tubes,
            'rating': self.rating,
            'is_locked': self.is_locked,
            'is_home_maze': self.is_home_maze,
            'is_published': self.is_published,
            'is_publish_expired': self.is_publish_expired,
            'player_id': self.player_id.to_dict(),
            'maze_pieces': [i.to_dict() for i in self.maze_pieces],
            'home_theme': self.home_theme.to_dict(),
            'parent_id': self.parent_id.to_dict(),
            'source_id': self.source_id.to_dict(),
        }

File: internal/game/types/player_maze.go
package types

import (
	"time"

	"github.com/dv1x3r/amazing-core/internal/game/gsf"
)

type PlayerMaze struct {
	OID              OID
	Name             string
	Size             int64
	Thumbnail        []byte
	PublishTimestamp time.Time
	NumRooms         int16
	NumTubes         int16
	Rating           gsf.Null[int16]
	IsLocked         bool
	IsHomeMaze       bool
	IsPublished      bool
	IsPublishExpired bool
	PlayerID         OID
	MazePieces       []PlayerMazePiece
	HomeTheme        AssetContainer
	ParentID         OID
	SourceID         OID
}

func (pm *PlayerMaze) Serialize(writer gsf.ProtocolWriter) {
	writer.WriteObject(&pm.OID)
	writer.WriteString(pm.Name)
	writer.WriteInt64(pm.Size)
	writer.WriteBytes(pm.Thumbnail)
	writer.WriteUtcDate(pm.PublishTimestamp)
	writer.WriteInt16(pm.NumRooms)
	writer.WriteInt16(pm.NumTubes)
	gsf.WriteNullable(writer, pm.Rating, writer.WriteInt16)
	writer.WriteBool(pm.IsLocked)
	writer.WriteBool(pm.IsHomeMaze)
	writer.WriteBool(pm.IsPublished)
	writer.WriteBool(pm.IsPublishExpired)
	writer.WriteObject(&pm.PlayerID)
	gsf.WriteSlice(writer, pm.MazePieces, func(value PlayerMazePiece) {
		writer.WriteObject(&value)
	})
	writer.WriteObject(&pm.HomeTheme)
	writer.WriteObject(&pm.ParentID)
	writer.WriteObject(&pm.SourceID)
}

func (pm *PlayerMaze) Deserialize(reader gsf.ProtocolReader) {
	reader.ReadObject(&pm.OID)
	pm.Name = reader.ReadString()
	pm.Size = reader.ReadInt64()
	pm.Thumbnail = reader.ReadBytes()
	pm.PublishTimestamp = reader.ReadUtcDate()
	pm.NumRooms = reader.ReadInt16()
	pm.NumTubes = reader.ReadInt16()
	pm.Rating = gsf.ReadNullable(reader, reader.ReadInt16)
	pm.IsLocked = reader.ReadBool()
	pm.IsHomeMaze = reader.ReadBool()
	pm.IsPublished = reader.ReadBool()
	pm.IsPublishExpired = reader.ReadBool()
	reader.ReadObject(&pm.PlayerID)
	pm.MazePieces = gsf.ReadSlice(reader, func() PlayerMazePiece {
		var value PlayerMazePiece
		reader.ReadObject(&value)
		return value
	})
	reader.ReadObject(&pm.HomeTheme)
	reader.ReadObject(&pm.ParentID)
	reader.ReadObject(&pm.SourceID)
}
